### PR TITLE
[MCC-651906] Add pluck method to Policy Machine and consume in at least one place in Dalton

### PR DESCRIPTION
**Pull request by okashansky-mdsol**
[MCC-651906] Add pluck method to Policy Machine and consume in at least one place in Dalton

Background
We need one of these for performance reasons, namely so that we can only select a single column or subset of columns from the database (instead of all the columns).

Changes Summary
The changes implemented and tested for  **PolicyMachineStorageAdapter::ActiveRecord** storage adapter. 
This is the PM-related part of the changes. Dalton-related changes are in Dalton repo.

* **pluck** method added to **PolicyMachineStorageAdapter::ActiveRecord**
* **pluck** method added to **PolicyMachine**
* specs added to **active_record_spec**

Scenarios
We can get an array of entity fields instead of ActiveRecord-descendant entity with code like:
```
<policy-machine-instance>.pluck(type: :type, fields: [<fields>], options: {<options> } )
```
or in Dalton (please use a branch in which implemented   ActivePolicyElement::Base.pluck)
```
<ActivePolicyElement::Base-descendant-class>.pluck(fields: [<fields>], options: {<options> } )
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,5 @@
 test/testapp/
 Gemfile.lock
 coverage/
+.idea/
+.vscode/
```
Comment (DMcKinnon-mdsol) [inline .gitignore:5]:
```suggestion
.vscode/

```

File: .idea/.gitignore [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# # Default ignored files
+# /shelf/
+# /workspace.xml
+# # Datasource local storage ignored files
+# /dataSources/
+# /dataSources.local.xml
+# # Editor-based HTTP Client requests
+# /httpRequests/
+.vscode/
+.idea/
```
Comment (DMcKinnon-mdsol) [inline .idea/.gitignore:10]: Is this file necessary?

File: lib/policy_machine.rb [hunk 3]
```diff
@@ -244,6 +244,15 @@ def batch_find(type:, query: {}, config: {}, &blk)
     end
   end
 
+  def pluck(type:, fields:, options:)
+    if policy_machine_storage_adapter.respond_to?(:pluck)
+      policy_machine_storage_adapter.send(:pluck, {type: type, fields: fields, options: options})
+    else
+      Warn.once("WARNING: 'pluck' is not implemented for storage adapter #{policy_machine_storage_adapter.class}")
+      nil
```
Comment (DMcKinnon-mdsol) [inline lib/policy_machine.rb:252]: I think we should actually hard raise here.

File: lib/policy_machine.rb [hunk 4]
```diff
@@ -244,6 +244,15 @@ def batch_find(type:, query: {}, config: {}, &blk)
     end
   end
 
+  def pluck(type:, fields:, options:)
+    if policy_machine_storage_adapter.respond_to?(:pluck)
+      policy_machine_storage_adapter.send(:pluck, {type: type, fields: fields, options: options})
```
Comment (DMcKinnon-mdsol) [inline lib/policy_machine.rb:249]:
```suggestion
      policy_machine_storage_adapter.send(:pluck, { type: type, fields: fields, options: options })
```

File: lib/policy_machine_storage_adapters/active_record.rb [hunk 5]
```diff
@@ -489,6 +489,12 @@ def operations
       end
     end # End of POLICY_ELEMENT_TYPES iteration
 
+    def pluck(type:, fields: [], options: {})
```
[the remark this answers is not in the document]
Comment (DMcKinnon-mdsol) [reply lib/policy_machine_storage_adapters/active_record.rb:492]: Oh, I see what you're doing now! Nevermind!

File: spec/policy_machine_storage_adapters/active_record_spec.rb [hunk 6]
```diff
@@ -92,7 +92,7 @@
             filters = { user_attributes: { color: 'purple' } }
 
             expect(
-              priv_pm.is_privilege_with_filters?(user_1, create, object_5, filters: filters)
+                priv_pm.is_privilege_with_filters?(user_1, create, object_5, filters: filters)
```
Comment (DMcKinnon-mdsol) [inline spec/policy_machine_storage_adapters/active_record_spec.rb:95]: Why did you make all these spacing changes? 2 spaces is the correct default.

File: spec/policy_machine_storage_adapters/active_record_spec.rb [hunk 7]
```diff
@@ -1697,6 +1697,17 @@
           end
         end
       end
+
+      describe 'pluck' do
+        (PolicyMachine::POLICY_ELEMENT_TYPES).each do |type|
+          it "can get data for #{type}" do
```
Comment (DMcKinnon-mdsol) [inline spec/policy_machine_storage_adapters/active_record_spec.rb:1703]:
```suggestion
          it "plucks the correct data for #{type}" do
```

File: spec/policy_machine_storage_adapters/active_record_spec.rb [hunk 8]
```diff
@@ -1697,6 +1697,17 @@
...
+            id = "#{type}-pluck-test"
+            policy_machine.send("create_#{type}", id)
+            data = policy_machine.send(:pluck, type: type, fields: [:unique_identifier], options: { unique_identifier: id })
+            expect(data[0]).to eq(id)
```
Comment (DMcKinnon-mdsol) [inline spec/policy_machine_storage_adapters/active_record_spec.rb:1707]: I would actually test the entire data structure like
```suggestion
            expect(data).to eq([id])
```

File: spec/policy_machine_storage_adapters/active_record_spec.rb (hunk 6 above)
Comment (okashansky-mdsol) [reply spec/policy_machine_storage_adapters/active_record_spec.rb:95]: I used to think that auto-formatting is basic thing of any development environment in XXI century. I was wrong...

File: spec/policy_machine_storage_adapters/active_record_spec.rb (hunk 8 above)
Comment (okashansky-mdsol) [reply spec/policy_machine_storage_adapters/active_record_spec.rb]: updated

File: .idea/.gitignore (hunk 2 above)
Comment (okashansky-mdsol) [reply .idea/.gitignore]: > Looking good! Mind also opening a pull request in Dalton that uses this branch and the pluck functionality?

I guess first we need to merge it to master? and create a tag. After that change Dalton's Gemfile probably.

File: lib/policy_machine.rb (hunk 3 above)
Comment (okashansky-mdsol) [reply lib/policy_machine.rb]: Now it throws the error.

File: lib/policy_machine.rb [hunk 9]
```diff
@@ -244,6 +244,14 @@ def batch_find(type:, query: {}, config: {}, &blk)
     end
   end
 
+  def pluck(type:, fields:, options:)
+    if policy_machine_storage_adapter.respond_to?(:pluck)
+      policy_machine_storage_adapter.send(:pluck, { type: type, fields: fields, options: options })
+    else
+      throw MethodNotSupportedError.new("Method 'pluck' is not implemented for storage adapter #{policy_machine_storage_adapter.class}")
```
Comment (DMcKinnon-mdsol) [inline lib/policy_machine.rb:251]: `throw` is extremely rare in Ruby. I've only actually seen it used in a handful of codebases. When it's used, iirc it should really only be used with catch. Otherwise we should probably just use `raise`. I think that's the case here. Let's just use `raise`.

**Comment by DMcKinnon-mdsol** [comment]
Please update the changelog file with a description of this change.

File: spec/policy_machine_storage_adapters/active_record_spec.rb [hunk 10]
```diff
@@ -1697,6 +1697,17 @@
           end
         end
       end
+
+      describe 'pluck' do
+        (PolicyMachine::POLICY_ELEMENT_TYPES).each do |type|
```
Comment (DMcKinnon-mdsol) [inline spec/policy_machine_storage_adapters/active_record_spec.rb:1702]: You don't need parenthesis here.

File: lib/policy_machine.rb (hunk 9 above)
Comment (okashansky-mdsol) [reply lib/policy_machine.rb]: fixed

File: spec/policy_machine_storage_adapters/active_record_spec.rb (hunk 10 above)
Comment (okashansky-mdsol) [reply spec/policy_machine_storage_adapters/active_record_spec.rb]: fixed
